### PR TITLE
adding support for nested structs

### PR DESCRIPTION
**Pull request by afsalthaj**
The change is made to support struct data type in thrifts.

### CODE REVIEW

**Comment by afsalthaj** [comment]
Build is failing saying:
`If you require sudo, add 'sudo: required' to your .travis.yml`

**Comment by afsalthaj** [comment]
@stephanh  Could you please look into this?

**Comment by laurencer** [comment]
The actual error is `The command "openssl aes-256-cbc -K $encrypted_d84550f15e95_key -iv $encrypted_d84550f15e95_iv -in .travis/deploy-key.enc -out .travis/deploy-key.pem -d" failed and exited with 1 during .` which means its failing to decrypt the deployment keys.

@stephanh - mind taking a look?

**Comment by afsalthaj** [comment]
@laurence. Does that mean I need to update the yml file with the latest encrypted keys?

**Comment by afsalthaj** [comment]
@laurence. Does that mean I need to update the yml file with the latest encrypted keys?

File: src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala [hunk 1]
```diff
@@ -149,4 +149,11 @@ object HiveMetadataTable {
 
   /** Gets the manifest for `T`. */
   def manifest[T : Manifest]: Manifest[T] = implicitly[Manifest[T]]
+
+  def getMetadataSchema(thrift: Class[_])= {
```
Comment (SamRoberts) [inline src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala]: I wrote this scala code some time ago for realtime. I'm not sure what `Reflect` does by comparison, but you may want to consider the "base trait" logic and see if that is relevant to your needs.

```scala
  /**
   * Find the ThriftCodec for a given thrift class.
   *
   * By convention, all ThriftStruct classes are created with a ThriftStructCodec
   * companion object. Annoyingly, however, the ThriftStruct constraint does not
   * allow us to access that object. For the moment, we are using unsafe runtime
   * reflection to access the companion object.
   */
  def getCodec(classWrapper: ThriftClass): UntypedCodec = {
    try {
      /* Scrooge generates a trait for each thrift struct, with a number
       * of class implementations. The trait has a corresponding object, as do
       * some of the classes. We want the object which corresponds to
       * the underlying trait, as this is the object which contains the
       * field information. klazz may either be the trait type or a class type.
       */
      val thriftClass = classOf[ThriftStruct]

      def getBaseTrait(klazz: Class[_]): Class[_] = {
        klazz.getInterfaces match {
          case Array(baseTrait) if baseTrait.getInterfaces contains (thriftClass) ⇒ baseTrait
          case interfaces if interfaces.contains(thriftClass)                     ⇒ klazz
          case _                                                                  ⇒ throw new IllegalArgumentException(s"unexpected thrift class hierarchy for $klazz")
        }
      }
      val traitClass = getBaseTrait(classWrapper.javaClass)
      val traitSyb = currentMirror.classSymbol(traitClass)
      val companionSyb = traitSyb.companion.asModule
      val companion = currentMirror.reflectModule(companionSyb).instance

      companion.asInstanceOf[UntypedCodec]
    } catch {
      case NonFatal(e) ⇒ throw new IllegalArgumentException(s"Cannot find ThriftStructCodec for ${classWrapper.javaClass}", e)
    }
  }
```

Comment (SamRoberts) [inline src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala]: return types for this method would come in handy

File: src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala [hunk 2]
```diff
@@ -40,13 +40,9 @@ object HiveMetadataTable {
     location: Option[Path] = None
   )(implicit m: Manifest[T]): MetadataTable = { // This operation could fail so type should convey it
     val thrift: Class[T]      = m.runtimeClass.asInstanceOf[Class[T]]
-    val codec                 = Reflect.companionOf(thrift).asInstanceOf[ThriftStructCodec[_ <: ThriftStruct]]
-    val metadata              = codec.metaData
-    //Making the types lowercase to enforce the same behaviour as in cascading-hive
-    val columnFieldSchemas    = metadata.fields.sortBy(_.id).map { c => 
-      fieldSchema(c.name, mapType(c).toLowerCase)
-    }
+    val (metadata, schema)    = getMetadataSchema(thrift)
```
Comment (SamRoberts) [inline src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala]: Here you refer to the second element in this pair as the `schema`. Further down the page you refer to the entire tuple as the `schema`. Would be good to be consistent.

File: src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala [hunk 3]
```diff
@@ -149,4 +149,11 @@ object HiveMetadataTable {
 
   /** Gets the manifest for `T`. */
   def manifest[T : Manifest]: Manifest[T] = implicitly[Manifest[T]]
+
+  def getMetadataSchema(thrift: Class[_])= {
+    val codec                 = Reflect.companionOf(thrift).asInstanceOf[ThriftStructCodec[_ <: ThriftStruct]]
+    val metadata              = codec.metaData
+    //Making the types lowercase to enforce the same behaviour as in cascading-hive
+    (metadata, metadata.fields.sortBy(_.id).map { c => (c.name, mapType(c).toLowerCase) })
```
Comment (SamRoberts) [inline src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala]: So I think that the second element of this pair is solely composed of data from `metadata`. In that case, you're not really returning any new information by returning this extra value. I would suggest just returning `metadata`, and having a separate function to convert from that `metadata` to these schema fields.

File: src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala [hunk 4]
```diff
@@ -106,7 +102,11 @@ object HiveMetadataTable {
       }
 
       // n type params
-      case TType.STRUCT => throw new Exception("STRUCT is not a supported Hive type")
+      case TType.STRUCT => {
+        val thrift = field.manifest.get.runtimeClass
+        val schema = getMetadataSchema(thrift)
+        s"struct<${schema._2.map(t => s"${t._1}:${t._2}").mkString(",")}>"
```
Comment (SamRoberts) [inline src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala]: I have a question about this. Wouldn't it be a great deal easier for users if you flattened the struct and moved all it's columns up into the top level of the table, with names like `outerField.nestedField1`, `outerField.nestedField2`, etc?

File: src/test/scala/au/com/cba/omnia/beeswax/HiveMetadataTableSpec.scala [hunk 5]
```diff
@@ -14,24 +14,31 @@
...
-  create a valid table for list                  $list
-  create a valid table for map                   $map
-  create a valid table for nested maps and lists $nested
+  create a valid  table for primitive types        $primitives
+  create a valid table for list                    $list
+  create a valid table for map                     $map
+  create a valid table for nested maps and lists   $nested
+  create a valid table for structs with primitives $structWithPrimitives
+  create a valid table for structs with map        $structWithMaps
+  create a valid table for structs with map        $structWithList
```
Comment (SamRoberts) [inline src/test/scala/au/com/cba/omnia/beeswax/HiveMetadataTableSpec.scala]: "structs with map" -> "structs with list"

**Comment by SamRoberts** [comment]
I don't really feel like I know enough about our current hive use to give an informed review, but for what it's worth, my review is done.

File: src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala (hunk 3 above)
Comment (afsalthaj) [inline src/main/scala/au/com/cba/omnia/beeswax/HiveMetadataTable.scala]: sure

**Comment by afsalthaj** [comment]
@triggerNZ  would u mind having a look at it?

**Pull request closed without merging**